apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,524 @@
+			defer wg.Done()
+			err := sci.limiter.Wait(context.Background())
+			if err != nil {
+				errChan <- fmt.Errorf("rate limiter error: %w", err)
+				return
+			}
+			sig, err := sci.pushLimitedBatchUpdateToContract(priceUpdateBatch)
+			if err != nil {
+				errChan <- fmt.Errorf("failed to push batch: %w", err)
+			} else {
+				sigChan <- sig
+			}
+		}(priceUpdateBatch)
+	}
+
+	wg.Wait()
+	close(errChan)
+	close(sigChan)
+	// Collect any errors
+	var errs []error
+	for err := range errChan {
+		errs = append(errs, err)
+	}
+
+	if len(errs) > 0 {
+		return fmt.Errorf("batch push encountered %d errors: %v", len(errs), errs)
+	}
+
+	sigs := []string{}
+	for sig := range sigChan {
+		sigs = append(sigs, sig.String())
+	}
+
+	sci.logger.Info().
+		Int("numUpdates", len(priceUpdates)).
+		Strs("batchTransactionSigs", sigs).
+		Msg("Successfully pushed batch updates to contract")
+
+	return nil
+}
+
+func (sci *SolanaContractInteracter) batchPriceUpdates(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) []map[InternalEncodedAssetId]AggregatedSignedPrice {
+	priceUpdatesBatches := []map[InternalEncodedAssetId]AggregatedSignedPrice{}
+
+	priceUpdatesBatch := make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+	i := 0
+	for encodedAssetId, priceUpdate := range priceUpdates {
+		priceUpdatesBatch[encodedAssetId] = priceUpdate
+		i++
+		if len(priceUpdatesBatch) == sci.batchSize || i == len(priceUpdates) {
+			batchCopy := make(map[InternalEncodedAssetId]AggregatedSignedPrice, len(priceUpdatesBatch))
+			for k, v := range priceUpdatesBatch {
+				batchCopy[k] = v
+			}
+			priceUpdatesBatches = append(priceUpdatesBatches, batchCopy)
+			priceUpdatesBatch = make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+		}
+	}
+
+	return priceUpdatesBatches
+}
+
+func (sci *SolanaContractInteracter) pushLimitedBatchUpdateToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) (solana.Signature, error) {
+
+	if len(priceUpdates) > MAX_BATCH_SIZE {
+		return solana.Signature{}, fmt.Errorf("Batch size exceeds limit, skipping update")
+	}
+
+	randomId, err := rand.Int(rand.Reader, big.NewInt(256))
+	if err != nil {
+		return solana.Signature{}, fmt.Errorf("failed to generate random ID: %w", err)
+	}
+	treasuryId := uint8(randomId.Uint64())
+
+	treasuryAccount := sci.treasuryAccounts[treasuryId]
+	instructions := []solana.Instruction{}
+	assetIds := []string{}
+
+	for encodedAssetId, priceUpdate := range priceUpdates {
+		var assetId [32]uint8
+		copy(assetId[:], encodedAssetId[:])
+		assetIds = append(assetIds, hex.EncodeToString(encodedAssetId[:]))
+
+		feedAccount := sci.feedAccounts[encodedAssetId]
+
+		//convert the quantized price to bin.Int128
+		quantizedPriceBigInt := new(big.Int)
+		quantizedPriceBigInt.SetString(string(priceUpdate.StorkSignedPrice.QuantizedPrice), 10)
+
+		quantizedPrice := bin.Int128{
+			Lo: quantizedPriceBigInt.Uint64(),
+			Hi: quantizedPriceBigInt.Rsh(quantizedPriceBigInt, 64).Uint64(),
+		}
```

Comment:
is the logic in this for loop tested? it's easy to mess things up when we're doing things like byte conversions - maybe we could pull out a `convertPriceUpdateToTemporalNumericValueEvmInput` helper method so we can unit test this

Reply:
Agree, will implement